src/js/RemotePlaybackPlugin.ts

```diff
@@ -0,0 +1,129 @@
+   return typeof window !== 'undefined' &&
+      'HTMLVideoElement' in window &&
+      'remote' in HTMLVideoElement.prototype;
+}
+
+export class RemotePlaybackPlugin extends Plugin {
+
+   public readonly log!: videojs.Log;
+
+   private readonly _player: VideoJsPlayer;
+   private readonly _airPlayManager: AirPlayManager;
+
+   public constructor(player: videojs.Player) {
+      super(player);
+      // The constructor receives a standard Video.js Player, but then we assert the type
+      // of our custom VideoJsPlayer. This is safe because the plugin has been registered
+      // at this point.
+      this._player = player as VideoJsPlayer;
+      this._airPlayManager = new AirPlayManager(this._player);
+      this.log('initializing Remote Playback plugin...');
+
+      this._initialize();
+   }
+
+   public dispose(): void {
+      super.dispose();
+      this.log('Remote Playback plugin disposed');
+   }
+
+   public get airPlayManager(): AirPlayManager {
+      return this._airPlayManager;
+   }
+
+   public getState(): RemotePlaybackState | null {
+      return this._airPlayManager.getState();
+   }
+
+   public isConnected(): boolean {
+      return this._airPlayManager.isConnected();
+   }
+
+   private _initialize(): void {
+      this.log('Initializing Remote Playback plugin...');
+      this._addAirPlayButton();
+      this._checkAvailability();
+   }
+
+   private _addAirPlayButton(): void {
+      this.log('Initializing AirPlay button...');
+
+      this._player.ready(() => {
+         this.log(LOG_MESSAGES.PLAYER_READY);
+         this._addButtonToControlBar();
+      });
+   }
+
+   private _addButtonToControlBar(): void {
+      const controlBar = this._player.getChild(COMPONENT_NAMES.CONTROL_BAR);
+
+      if (!controlBar) {
+         this.log.error(LOG_MESSAGES.CONTROL_BAR_NOT_FOUND);
+         return;
+      }
+
+      try {
+         const airPlayButton = new AirPlayButton(this._player, { addAirPlayLabelToButton: true });
+
+         this.log(LOG_MESSAGES.BUTTON_CREATED);
+
+         this._insertButtonInControlBar(controlBar, airPlayButton);
+         this.log(LOG_MESSAGES.BUTTON_ADDED);
+      } catch(error) {
+         this.log.error('Failed to create or add AirPlay button', error);
+      }
+   }
+
+   private _insertButtonInControlBar(controlBar: unknown, airPlayButton: AirPlayButton): void {
+      const fullscreenToggle = (controlBar as { getChild: (name: string) => unknown }).getChild(COMPONENT_NAMES.FULLSCREEN_TOGGLE);
+
+      if (fullscreenToggle) {
+         const children = (controlBar as { children: () => unknown[] }).children();
+
+         const insertIndex = children.indexOf(fullscreenToggle);
+
+         (controlBar as { addChild: (component: unknown, options?: Record<string, unknown>, index?: number) => void })
+            .addChild(airPlayButton, {}, insertIndex);
+      } else {
+         (controlBar as { addChild: (component: unknown) => void }).addChild(airPlayButton);
+      }
+   }
```

Comment:
This typing must be improved! Here is a suggestion on how to improve it if it were to stay a separate function, and the key is requiring `controlBar` to be a `videojs.Component`.  But if this code is moved into `_addButtonToController` as suggested above, you'd have the right typing anyway because `this._player.getChild` is already sending you `controlBar` as a `videojs.Component`. 


```suggestion
   private _insertButtonInControlBar(controlBar: videojs.Component, airPlayButton: AirPlayButton): void {
      const fullscreenToggle = controlBar.getChild(COMPONENT_NAMES.FULLSCREEN_TOGGLE);

      // If there is a fullscreen toggle, insert the AirPlay button before it. Otherwise,
      // just add it to the end of the control bar.
      if (fullscreenToggle) {
         const children = controlBar.children(),
               fullscreenToggleIndex = children.indexOf(fullscreenToggle),
               insertIndex = fullscreenToggleIndex >= 0 ? fullscreenToggleIndex : children.length;

         controlBar.addChild(airPlayButton, {}, insertIndex);
      } else {
         controlBar.addChild(airPlayButton);
      }
   }
```